File: app/src/main/java/com/cliniva/enventory/adapter/base/BaseRecyclerViewHolder.java
package com.cliniva.enventory.adapter.base;

import android.content.Context;
import android.databinding.ViewDataBinding;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.cliniva.enventory.utils.OnSingleClickListener;

/**
 * Created by taohid on 19,February, 2019
 * Email: taohid32@gmail.com
 */
public abstract class BaseRecyclerViewHolder<T, L extends BaseRecyclerClickListener> extends RecyclerView.ViewHolder {

    public Context context;

    public BaseRecyclerViewHolder(@NonNull ViewDataBinding itemView) {
        super(itemView.getRoot());
        this.context = itemView.getRoot().getContext();
    }

    public abstract void onBindView(T item, int position, L listener);

    public void enableItemViewClick(T item, BaseRecyclerClickListener<T> baseRecyclerClickListener){

        itemView.setOnClickListener(new OnSingleClickListener() {

            @Override
            public void onSingleClick(View v) {

                if (baseRecyclerClickListener != null){
                    baseRecyclerClickListener.onItemClickListener(item, getAdapterPosition());
                }
            }
        });
    }
}

File: app/src/main/java/com/amirsons/inventory/recyclerview/base/BaseRecyclerViewHolder.kt
package com.amirsons.inventory.recyclerview.base

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.amirsons.inventory.utils.OnSingleClickListener

/**
 * Created by taohid on 19,February, 2019
 * Email: taohid32@gmail.com
 */

abstract class BaseRecyclerViewHolder<T, L : BaseRecyclerClickListener<T>>(itemView: View) : RecyclerView.ViewHolder(itemView) {

    var context: Context = itemView.context

    abstract fun onBindView(item: T, position: Int, listener: L?)

    fun enableItemViewClick(item: T, baseRecyclerClickListener: BaseRecyclerClickListener<T>?) {

        itemView.setOnClickListener(object : OnSingleClickListener() {

            override fun onSingleClick(v: View) {
                baseRecyclerClickListener?.onItemClickListener(item, adapterPosition)
            }
        })
    }
}
